crates/rpc/src/cairo/starknet_rs/state_reader.rs

```diff
@@ -0,0 +1,244 @@
+    pub storage: pathfinder_storage::Storage,
+    pub block_number: Option<BlockNumber>,
+    // Classes in pending state have already been downloaded and added to the database.
+    // This flag makes it possible to find these classes -- essentially makes the state
+    // reader look up classes which are not declared at a canonical block yet.
+    pub ignore_block_number_for_classes: bool,
+}
+
+impl PathfinderStateReader {
+    fn state_block_id(&self) -> Option<pathfinder_storage::BlockId> {
+        self.block_number.map(Into::into)
+    }
+}
+
+impl StateReader for PathfinderStateReader {
+    fn get_class_hash_at(
+        &self,
+        contract_address: &starknet_in_rust::utils::Address,
+    ) -> Result<starknet_in_rust::utils::ClassHash, StateError> {
+        let pathfinder_contract_address = pathfinder_common::ContractAddress::new_or_panic(
+            Felt::from_be_slice(&contract_address.0.to_bytes_be())
+                .expect("Overflow in contract address"),
+        );
+
+        let _span = tracing::debug_span!("get_class_hash_at", contract_address=%pathfinder_contract_address).entered();
+
+        tracing::trace!("Getting class hash at contract");
+
+        let block_id = self
+            .state_block_id()
+            .ok_or_else(|| StateError::NoneClassHash(contract_address.clone()))?;
+
+        let mut db = self.storage.connection().map_err(map_anyhow_to_state_err)?;
+        let tx = db.transaction().map_err(map_anyhow_to_state_err)?;
+
+        let class_hash = tx
+            .contract_class_hash(block_id, pathfinder_contract_address)
+            .map_err(|error| {
+                tracing::error!(%error, "Failed to fetch contract class hash");
+                StateError::CustomError(format!(
+                    "Failed to fetch contract class hash for contract {}",
+                    pathfinder_contract_address
+                ))
+            })?
+            .ok_or_else(|| StateError::NoneClassHash(contract_address.clone()))?;
+
+        Ok(class_hash.0.to_be_bytes())
+    }
+
+    fn get_nonce_at(
+        &self,
+        contract_address: &starknet_in_rust::utils::Address,
+    ) -> Result<Felt252, starknet_in_rust::core::errors::state_errors::StateError> {
+        let pathfinder_contract_address = pathfinder_common::ContractAddress::new_or_panic(
+            Felt::from_be_slice(&contract_address.0.to_bytes_be())
+                .expect("Overflow in contract address"),
+        );
+
+        let _span =
+            tracing::debug_span!("get_nonce_at", contract_address=%pathfinder_contract_address)
+                .entered();
+
+        tracing::trace!("Getting nonce for contract");
+
+        let block_id = self
+            .state_block_id()
+            .ok_or_else(|| StateError::NoneNonce(contract_address.clone()))?;
+
+        let mut db = self.storage.connection().map_err(map_anyhow_to_state_err)?;
+        let tx = db.transaction().map_err(map_anyhow_to_state_err)?;
+
```

Comment:
We're going to run into db sync issues like this. Ideally a single db tx should be opened for the entirety of the execution?

Its unlikely at the moment since we never delete history, but a reorg could invalidate data here. And much more likely, in the future you could be executing and then we delete the state for that block for non-archive nodes.

Reply:
Not sure how viable that is, or how much more complicated that becomes? This is probably okay for a 1st impl, but is a problem long term.

Reply:
Not so complicated, so I've changed `PathfinderStateReader` to use a single transaction instead.